Symplify autowiring and remove backticks

parijke opened the pull request:
- adhere to the most recent best practices
- remove ORM dependency (using DBAL) 
- simplify autowiring and service definitions
- remove backticks from select statements
- using 6.1 bundle definition
- moved services and routes from src/Resources/config to ./config

fixes #15 

Advantages: really simple services.yml. If you want to override the autowiring defaults, you could define an entry in the services.yaml.

How it works:
All services which implements `HealthCheckInterface` are tagged with `openconext.monitor.health_check`, that is why is is not needed in the Extension anymore. The TaggableIterator is initialized with al services tagged like this. That is why the CompilerPass becomes obsolete.

Just as flexible as before, just much simpeler. Thanks Symfony!
Also see: https://symfony.com/doc/current/bundles/extension.html

parijke committed:
Symplify autowiring and remove backticks

- remove ORM dependency

parijke committed:
Fix tests using iterator

parijke committed:
Update composer.json

parijke committed:
Remove obsolete service definitions

parijke committed:
Symplify checks because they are already iterable

parijke committed:
Simplify bundle config

parijke committed:
Move services to main config

parijke committed:
Move routes to main config

parijke committed:
Refactor services to new filestructure

parijke committed:
Adjust file location

parijke committed:
Add / adjust instructions

parijke committed:
Add AutoConfigureTag

parijke committed:
Remove 5.4 as supported version

parijke committed:
Remove tag name and use the interface

MKodde commented on README.md:
````diff
@@ -104,35 +98,20 @@ class ApiHealthCheck implements HealthCheckInterface
-            - { name: surfnet.monitor.health_check }
-```
+By implementing the `HealthCheckInterface` you can register your own health check. 
+This interface is tagged with `openconext.monitor.health_check` so you don't have to do it yourself.
````
Now this is what I call progress!

MKodde replied:
One thought tho. Quite some people now use this old construct. Have you verified that stuff keeps working without them changing any code? If a change is needed, this PR should trigger a new major release as we are breaking BC. Also in that case. A small readme on how to migrate to the new version is a must.

But as I see it from my code review chair, I think your change does not cause a BC break. But please check this!

parijke replied:
The only thing is the path to the routes. They are changed from Resources to /config

We can mention that in the release notes?

MKodde replied:
Rule of thumb; if any changes need to be done by the consumer due to a change made in a new release. We should provide an `upgrading.md` entry. Describing the way you can port from the previous version to the new way of doing things.

If there is no BC breakage, there is no need to document it.

MKodde commented on src/Controller/InfoController.php:
```diff
@@ -43,15 +42,23 @@
 
     public function __construct(
-        private readonly string $buildPath,
+        #[Autowire(param: 'kernel.project_dir')]
```
It still takes some getting used to. My personal pref still is using binded named values in the services.yaml. But this solution scores low on obscurity.

parijke replied:
And it is the way of the future. You can still overwrite this in the services.yaml if you need to...

parijke replied:
I also think it needs some time to get used to it. I felt exactly the same when I first saw attrutes at all.

MKodde commented on src/HealthCheck/DoctrineConnectionHealthCheck.php:
```diff
@@ -30,34 +31,28 @@
 
     public function check(HealthReportInterface $report): HealthReportInterface
     {
         // Was the entityManager injected? When it is not the project does not use Doctrine ORM
```
1. Please update the doc block here. This might cause confusion in the future as ORM is no longer used. 
3. Also, maybe test in the constructor if the service was set correctly? That way the app halts at an earlier stage, and your control logic is cleaner

parijke replied:
Very good point

parijke replied:
Hmmm, what should be the behaviour then if the connection is not there at constructor time? We cannot return $report then, like it is now...

parijke replied:
Not sure how to move the connection check to the constructor and preserve behaviour? Any suggestions? Docblock removed, it is self explaning code IMHO

MKodde replied:
You are right we explicitly allow for the service to be null. No use to check it in the constructor then..

Leave it like it is right now!

MKodde commented on src/HealthCheck/DoctrineConnectionHealthCheck.php:
```diff
@@ -30,34 +31,28 @@
             try {
                 // Get the schema manager and grab the first table to later query on
-                $sm = $this->entityManager->getConnection()->createSchemaManager();
+                $sm = $this->connection->createSchemaManager();
```
:heart_eyes_cat:

MKodde commented on src/HealthCheck/HealthCheckInterface.php:
```diff
@@ -18,9 +18,12 @@
 /**
  * Contract for a HealthCheck.
  */
+#[AutoconfigureTag()]
```
I feel this should be `#[AutoconfigureTag('surfnet.monitor.health_check')]`

parijke replied:
Can be, but if you skip the tag, Symfony will use the FQN of the interface (I learned  yesterday). Awesome

bin/console deb:cont --tag="OpenConext\MonitorBundle\HealthCheck\HealthCheckInterface"

Symfony Container Services Tagged with "OpenConext\MonitorBundle\HealthCheck\HealthCheckInterface" Tag

 -------------------------------------------------------------------- -------------------------------------------------------------------- 
Service ID                                                           Class name
 -------------------------------------------------------------------- -------------------------------------------------------------------- 
OpenConext\MonitorBundle\HealthCheck\DoctrineConnectionHealthCheck   OpenConext\MonitorBundle\HealthCheck\DoctrineConnectionHealthCheck  
OpenConext\MonitorBundle\HealthCheck\SessionHealthCheck              OpenConext\MonitorBundle\HealthCheck\SessionHealthCheck
 -------------------------------------------------------------------- --------------------------------------------------------------------

MKodde replied:
In that case, you should update the readme and state that those services are tagged with the FQN of the interface. And that is a BC break. Now upgrading users need to remove their userland tagged services and solely rely on the SF auto-tag logic.

parijke replied:
That actually is a good argument I did not think of. In that case, we'll re-introduce the tag, as it works fine with.

Only the code was a bit clearer to me just using the interface. But it is a small trade off for keeping BC

parijke replied:
Reverted back to using tags

MKodde commented on src/OpenConextMonitorBundle.php:
```diff
@@ -18,14 +18,14 @@
+    public function loadExtension(array $config, ContainerConfigurator $container, ContainerBuilder $builder): void
     {
-        $container->addCompilerPass(new HealthCheckPass());
+        $container->import('../config/services.yaml');
```
Safe to use a relative path here? Maybe use a param to refer to the config?

`kernel.project_dir`

parijke replied:
It is how its done in the Symfony Documentation, so pretty safe I would say
https://symfony.com/doc/current/bundles/extension.html#loading-services-directly-in-your-bundle-class

parijke replied:
BTW, kernel.projectDir refers to the consuming project. The path here is relative to the bundle (hence the move from Resources to the /config of the bundle structure

![image](https://github.com/OpenConext/Monitor-bundle/assets/7671998/3208b904-ab86-4370-bead-6708dc5043a4)

MKodde replied:
Thanks for your explanation

MKodde reviewed (commented):
A very nice effort to actually lever some useful newer Symfony constructions into the project. And a good call to stop using Doctrine ORM but opt for DBAL instead :tophat: 

I have some minor concerns listed below. Please have at them!

MKodde commented:
@tacman does this solve your issues as proposed in #15? The other suggestion to allow for setting a custom query is still in our pipeline. We recently ran into an issue where checking existence of a certain table was very useful.

parijke committed:
Solved "error"?

I am unsure if it was an error as the test results do not change after this fix. This is a smell of an untrusted test

parijke commented on src/Tests/HealthCheck/SessionHealthCheckChainTest.php:
```diff
@@ -73,7 +74,7 @@ public function testChainStopsWhenCheckFails()
             ->andReturn(false);
 
         $statusDown = m::mock(HealthReport::buildStatusDown('Lorem ipsum dolor sit'));
-        $statusOk
```
@MKodde can you confirm this was an mistake?

The problem is, in both cases the test passes

MKodde replied:
Good catch. That looks like an error on my account. I do understand that the test passed. In this case a spy would have been usefull to later check if the second checker caused the downage.

But thanks for the fix!

parijke committed:
Remove unused dependencies

parijke committed:
Adjust description

parijke committed:
Move tests out of src

parijke committed:
Remove extra character

parijke committed:
Remove obsolete settings

parijke committed:
Fix deprecation

(see https://github.com/symfony/symfony/commit/4bc6bedfdb21f8de6cfcfd9860aa10490a321b95)

MKodde reviewed (commented):
One todo remains. I think adding the service tag is important to help existing users to upgrade with `NOOP`.

parijke committed:
Switch to attribute routes

parijke committed:
Add correct routes to config installer

parijke committed:
Update with routing

parijke committed:
Add information about the routin installer

parijke committed:
Re-anabled the tags

parijke reviewed (commented):
Reverted back to using tags
